### ISSUE TITLE
Code Coverage Tool

### ISSUE THREAD

**Issue by apcraig** [CICE-Consortium/CICE#454]
This is a longish issue, sorry about that.  I've tried to summarize the current situation.  Looking for input from @eclare108213 @duvivier @rallard77 @phil-blain @dabail10 and others.

I am having problems with codecov again.  It was working fine a couple weeks ago, but with the increased number of tests we're now running on cheyenne (due to pio and IO testing in general), the codecov reports are NOT working again.  I have created a couple of codecov issues, but am not expecting things to improve.  I don't know how much longer we should fight this battle.

In the last couple days, I have been playing with lcov.  lcov is an older perl script that analyzes the same gnu gcov output.  It comes with a tool that produces html output and I have verfied that I can post the html at github thru the github.io feature.  Some sample output is here,

https://apcraig.github.io/25dd6d76f4_t1/index.html

and for comparison, similar codecov output is here,

https://codecov.io/gh/apcraig/Test_CICE_Icepack/list/3d1a95bc42877d9ab1a5ddf17d5dd255ce4d929b/

To summarize the two approaches:

With codecov, we compile with the gnu compiler and special compiler flags.  When all runs are done, we download the codecov bash script which then processes the model output with gcov (part of the gnu compiler) into gcov output files.  Those results are concatenated and the resulting file is sent to the codecov.io website, aggregated, and posted.

With lcov, we compile with the gnu compiler and the same special compiler flags.  We have to install lcov and some supporting software one-time.  Then we run lcov on the model output which is processed with gcov into lcov output files.  We aggregate the lcov output files with lcov and then run another tool to generate an html directory.  That html directory can then be posted to the web, like at github.io relatively easily.

There are pluses and minuses to codecov vs lcov.  The most important one is that lcov seems to be robust while codecov is not.  But let me try to list the differences.

- Robustness: lcov wins out at this point.  codecov has been extremely unreliable.
- Output: Both lcov and codecov show overall statistics as well as file-by-file and line-by-line results.  They are very similar as they use the same gnu/gcov output for processing.
- Website output: codecov is a lot nicer!
- Software: both codecov and lcov leverage the same gnu/gcov tools for initial processing of runs on the client side.  Different processes are used to aggregate and post the results.
- Installation/Usage: I had to manually install lcov and some supporting software on cheyenne.  It was not difficult, but would probably need to be done by anyone using lcov, but we can probably install once per machine and share.  With codecov, there is no special installation required.  A bash script is downloaded from the codecov site to your local directory and then the processing is done with that.
- Multiple Users: With codecov, we need to share the token, but then it's relatively easy for multiple users to test.  With lcov, we would need to add write permission to the github.io repository to users that want to test so they can post results.  There is also the installation issue above.
- Testing: With codecov, we can test on multiple machines and with different suites for a given model version and those results are automatically aggregated on the codecov site.  Although this is one of the features that has not been very robust.  With lcov, it's most convenient to run a full test suite once to produce full results.  It is possible to run multiple suites on multiple machines with lcov, but we'd have to manually pull the test results together onto one machine to generate a complete set of html output at once.  At the present time, our full test suite only runs on cheyenne (because of pio), so it's unclear how much flexibility we need anyway.
- Large Testsuite: codecov has problems both posting results from a large test suite and aggregating multiple results from smaller test suites.  lcov handles our large test suite fine.
- Performance: both codecov and lcov use gcov for processing the model output and gcov is not fast.  It takes about 2 minutes per test to process data on cheyenne and the full test suite on cheyenne is 137 tests right now.  So that's 4+ hours for the basic gcov analysis.  However, the aggregation step is done differently which takes lcov about 10 min and codecov about 6 hours, so lcov is faster than codecov by about 2x overall.  Both tools takes some time to analyze the full suite of output.
- Control/Debugging: lcov wins out because the codecov bash script has a number of very finicky features about how files need to be organized to be processed and many other things.  Also, the codecov bash script is a bit of a black box and debugging is difficult.  lcov is run on the command line and provides adequate flexibility and is pretty easy to use.
- Risk: The codecov bash script is likely to evolve in time which introduces risk in the processing part.  We also have no control on the server side (codecov website) when things break.  With lcov, we control the lcov version, lcov update cycle, as well as the website.
- Badge: codecov provides one.  We could probably create one ourselves pretty easily for our lcov output.

I think the question is how to proceed.  At this point, I think lcov is the best way forward.  If the tool is going to be used by the internal group occasionally, and what we mostly want is to diagnose coverage and improve the test suite, lcov seems more than adequate and is robust.  We do not have to abandon codecov.  We can keep the current scripts in place and continue to test it.  I think that's also a good idea.  So this is what I propose

- Put in place the lcov capability and use this by default to assess coverage
- Add some documentation about lcov as well as how to install it
- Decide where to store the lcov output.  I am happy to volunteer apcraig.github.io and will assume we'll use that for now.  We could also create a cice-consortium.github.io repository to store the results.  I think we should store the last few results (assuming we test a few times a year), but can clean up older results occasionally manually.  We use a similar process for the test results.
- Plan to test with lcov by running the full test suite on cheyenne periodically.  This somewhat limits our testing flexibility in terms of people and machines, but it's a reasonable starting point given that the full test suite can only be run on cheyenne right now (due to pio).  We can look into installing pio on other machines in the future.
- Maintain the codecov capability and test occasionally.

**Comment by eclare108213** [comment]
Thanks @apcraig for the thorough comparison!  I'm glad to know there is a reasonable alternative to codecov.  Your suggestions above sound fine, although I question maintaining both lcov and codecov.   It makes sense to run this each time we do a new release, or when we add a new test, and store the results in a consortium web space.  And for this purpose, just running on one machine seems adequate.

**Comment by apcraig** [comment]
lcov has been implemented in both CICE and Icepack in #467 and https://github.com/CICE-Consortium/Icepack/pull/324.  We have also decided to temporarily abandon codecov, but the capability to assess coverage with it was not removed from the CICE or Icepack scripts.  Documentation is update to reflect use of lcov and a new wiki page has been added, https://github.com/CICE-Consortium/About-Us/wiki/LCOV-Installation, to document lcov installation.
